### PR TITLE
dataclasses: various fixes

### PR DESCRIPTION
**Pull request by hauntsaninja**

### CODE REVIEW

File: stdlib/3.7/dataclasses.pyi [hunk 1]
```diff
@@ -18,9 +18,9 @@ def astuple(obj: Any) -> Tuple[Any, ...]: ...
 @overload
 def astuple(obj: Any, *, tuple_factory: Callable[[List[Any]], _T]) -> _T: ...
 @overload
-def dataclass(_cls: Type[_T]) -> Type[_T]: ...
+def dataclass(__cls: Type[_T]) -> Type[_T]: ...
 @overload
-def dataclass(_cls: None) -> Callable[[Type[_T]], Type[_T]]: ...
+def dataclass(__cls: None) -> Callable[[Type[_T]], Type[_T]]: ...
```
Comment (srittau) [inline stdlib/3.7/dataclasses.pyi:23]: The stubtest complaints are valid. The following is horrible, but possible:

```python
from dataclasses import dataclass

class Foo:
    x: int
    y: str
dataclass(_cls=Foo)

print(Foo(42, ""))
```

File: stdlib/3.7/dataclasses.pyi [hunk 2]
```diff
@@ -76,6 +98,8 @@ def is_dataclass(obj: Any) -> bool: ...
 class FrozenInstanceError(AttributeError): ...
 
 class InitVar(Generic[_T]):
```
Comment (JelleZijlstra) [inline stdlib/3.7/dataclasses.pyi:100]: Why is this Generic? Should `type` be of type `Type[_T]`?

File: stdlib/3.7/dataclasses.pyi [hunk 3]
```diff
@@ -17,14 +17,26 @@ def asdict(obj: Any, *, dict_factory: Callable[[List[Tuple[str, Any]]], _T]) ->
...
-@overload
-def dataclass(_cls: None) -> Callable[[Type[_T]], Type[_T]]: ...
-@overload
-def dataclass(
-    *, init: bool = ..., repr: bool = ..., eq: bool = ..., order: bool = ..., unsafe_hash: bool = ..., frozen: bool = ...
-) -> Callable[[Type[_T]], Type[_T]]: ...
+
+if sys.version_info >= (3, 8):
+    @overload
+    def dataclass(__cls: Type[_T]) -> Type[_T]: ...
```
Comment (JelleZijlstra) [inline stdlib/3.7/dataclasses.pyi:23]: Took me a long time to figure out that the difference between the branches is `__cls` vs. `_cls`. Maybe it's worth a comment.

**Comment by hauntsaninja** [comment]
Thanks, addressed (also fixed up `__class_getitem__` since it doesn't return GenericAlias)

**Pull request merged**